contracts/staking/DelegationManager.sol

```diff
@@ -1,230 +1,274 @@
+  }
+
+  function unBond(uint256 delegatorId) public onlyDelegator(delegatorId) {
+    // TODO: validator amount update
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
+    // _claimRewards(delegatorId);
+    uint256 currentEpoch = stakeManager.currentEpoch();
+    _unbond(delegatorId, currentEpoch, stakeManager);
+    emit UnBonding(delegatorId, delegators[delegatorId].bondedTo);
+  }
+
+  function _unbond(uint256 delegatorId, uint256 epoch,  IStakeManager stakeManager) private {
+    stakeManager.updateValidatorState(delegators[delegatorId].bondedTo, epoch, -int(delegators[delegatorId].amount));
+    validatorDelegation[delegators[delegatorId].bondedTo] = validatorDelegation[delegators[delegatorId].bondedTo].sub(delegators[delegatorId].amount);
+    delegators[delegatorId].bondedTo = 0;
+  }
+
+  function reStake(uint256 delegatorId, uint256 amount, bool stakeRewards) public onlyDelegator(delegatorId) {
+    Delegator storage delegator = delegators[delegatorId];
+    if (amount > 0) {
+      require(token.transferFrom(msg.sender, address(this), amount), "Transfer stake");
+    }
+    if (stakeRewards) {
+      amount += delegator.reward;
+      delegator.reward = 0;
+    }
+    totalStaked = totalStaked.add(amount);
+    if (delegator.bondedTo != 0) {
+      validatorDelegation[delegators[delegatorId].bondedTo] = validatorDelegation[delegators[delegatorId].bondedTo].add(delegator.amount);
+    }
+
+    delegator.amount = delegator.amount.add(amount);
+    emit ReStaked(delegatorId, amount, totalStaked);
+  }
+
+  function slash(uint256[] memory _delegators, uint256 slashRate) public  {
+      // Validate
+      // for (uint256 i; i < _delegators.length; i++) {
+      //   Delegator storage delegator = delegators[_delegators[i]];
+      //   delegator.amount = delegator.amount.sub(delegator.amount.mul(slashRate).div(100));
+      // }
+      // uint256 slashedAmount = 0
+      // validatorDelegation[validatorId] = validatorDelegation[validatorId].sub(amount);
+  }
+
+  function claimRewards(
+    uint256 checkpointId,// checkpoint Id  with root of proofs
+    uint256 delegatorId,
+    uint256 rewardAmount,
+    uint256 slashedAmount,
+    uint256 accIndex,
+    bool withdraw,
+    bytes memory accProof
+    ) public isDelegator(delegatorId) /*onlyDelegator(delegatorId) */ {
+    IStakeManager stakeManager = IStakeManager(registry.getStakeManagerAddress());
+    require(
+      keccak256(
+        abi.encodePacked(
+          delegatorId,
+          rewardAmount,
+          slashedAmount)
+          ).checkMembership(
+            accIndex,
+            accRoot[checkpointId],
+            accProof),
+      "Wrong account proof"
+      );
+
+    uint256 _rewardAmount = rewardAmount.sub(delegators[delegatorId].claimedRewards);
+    if (_rewardAmount <= slashedAmount) {
+      delegators[delegatorId].amount = delegators[delegatorId].amount.sub(slashedAmount.sub(_rewardAmount));
```

Comment:
only delegator amount is reduced. TotalSupply also should be reduced.

Reply:
Also validatorDelegation should be reduced i think

Reply:
fixed.